test/e2e/pkg/grammar/checker_test.go

```diff
@@ -0,0 +1,103 @@
+	// state-sync = success-sync
+	{"one-apply-chunk-1", []*abci.Request{initChain, offerSnapshot, applyChunk, finalizeBlock, commit}, CLEAN_START, PASS},
+	{"multiple-apply-chunks-1", []*abci.Request{initChain, offerSnapshot, applyChunk, applyChunk, finalizeBlock, commit}, CLEAN_START, PASS},
+	{"offer-snapshot-missing-1", []*abci.Request{initChain, applyChunk, finalizeBlock, commit}, CLEAN_START, FAIL},
+	{"apply-chunk-missing", []*abci.Request{initChain, offerSnapshot, finalizeBlock, commit}, CLEAN_START, FAIL},
+	// state-sync = *state-sync-attempt success-sync
+	{"one-apply-chunk-2", []*abci.Request{initChain, offerSnapshot, applyChunk, offerSnapshot, applyChunk, finalizeBlock, commit}, CLEAN_START, PASS},
+	{"mutliple-apply-chunks-2", []*abci.Request{initChain, offerSnapshot, applyChunk, applyChunk, applyChunk, offerSnapshot, applyChunk, finalizeBlock, commit}, CLEAN_START, PASS},
+	{"offer-snapshot-missing-2", []*abci.Request{initChain, applyChunk, offerSnapshot, applyChunk, finalizeBlock, commit}, CLEAN_START, FAIL},
+	{"no-apply-chunk", []*abci.Request{initChain, offerSnapshot, offerSnapshot, applyChunk, finalizeBlock, commit}, CLEAN_START, PASS},
+
+	// start = recovery
+	// recovery = consensus-exec
+	// consensus-height = finalizeBlock commit
+	{"empty-block-2", []*abci.Request{finalizeBlock, commit}, !CLEAN_START, PASS},
+	{"finalize-block-missing-2", []*abci.Request{commit}, !CLEAN_START, FAIL},
+	{"commit-missing-2", []*abci.Request{finalizeBlock}, !CLEAN_START, FAIL},
+	// consensus-height = *consensus-round finalizeBlock commit
+	{"proposer-round-3", []*abci.Request{prepareProposal, processProposal, finalizeBlock, commit}, !CLEAN_START, PASS},
+	{"proposer-round-4", []*abci.Request{prepareProposal, finalizeBlock, commit}, !CLEAN_START, PASS},
+	{"non-proposer-round-2", []*abci.Request{processProposal, finalizeBlock, commit}, !CLEAN_START, PASS},
+	{"multiple-rounds-2", []*abci.Request{prepareProposal, processProposal, processProposal, prepareProposal, processProposal, processProposal, processProposal, finalizeBlock, commit}, !CLEAN_START, PASS},
+
+	// corner cases
+	{"empty execution", nil, CLEAN_START, FAIL},
+	{"empty execution", nil, !CLEAN_START, FAIL},
+}
+
+func TestVerify(t *testing.T) {
+	for _, test := range tests {
+		checker := NewGrammarChecker(DefaultConfig())
+		result, err := checker.Verify(test.abciCalls, test.isCleanStart)
+		if result == test.result {
+			continue
+		}
+		if err == nil {
+			err = fmt.Errorf("Grammar parsed an incorrect execution: %v\n", checker.getExecutionString(test.abciCalls))
+		}
+		t.Errorf("Test %v returned %v, expected %v\n%v\n", test.name, result, test.result, err)
+	}
+}
+
+func TestFilterLastHeight(t *testing.T) {
+	reqs := []*abci.Request{initChain, finalizeBlock, commit}
+	checker := NewGrammarChecker(DefaultConfig())
+	rr, n := checker.filterLastHeight(reqs)
+	if len(reqs) != len(rr) || n != 0 {
+		t.Errorf("FilterLastHeight filtered %v ABCI calls, expected %v\n", n, 0)
+	}
```

Comment:
```suggestion
	require.Equal(t, len(reqs),len(rr), "FilterLastHeight check failed with filtered ABCI calls")
	require.Zero(t, n, "Check failed with filtered ABCI calls")
```
Please use this pattern all over your unit tests (actually, we usually don't need the message parameter, as the framework will always show the line where the failing check was, and the values that failed the check)